Add delayed upload mode in MathBackendWebGL

dsmilkov opened the pull request:
- Add delayed upload mode in `MathBackendWebGL`. When delayed mode is `true`, newly constructed `NDArray`s won't get immediately uploaded to the GPU. Also when user reads back the values, the backing texture gets disposed to reduce probability of memory leaks. Delayed mode is `true` by default.
- Fix `math.getNumArrays()`
- Fix memory leaks in Squeezenet
- Fix bugs in `models/*/package.json` (`tsc` creates different dir structure under `dist/` once we added dependency on `../util.ts`)
- Fix imagenet demo flickering

After this PR is merged, I'll release new npm versions in this order:
1) deeplearn
2) deeplearn-squeezenet
3) deeplearn-knn-image-classifier

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pair-code/deeplearnjs/466)
<!-- Reviewable:end -->

nsthorat commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

Nice work!

---

Reviewed 14 of 14 files at r1.
Review status: 13 of 14 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[demos/imagenet/imagenet.ts, line 238 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/466#-L1G3ayq7VfQwzyHRyMv:-L1G3ayq7VfQwzyHRyMw:b-ts8svg) ([raw file](https://github.com/pair-code/deeplearnjs/blob/4249b0e0c75ebc1a3c8483a3f24b30713b1ddfbc/demos/imagenet/imagenet.ts#L238)):*
> ```XML
>           activationNDArray.shape[0], activationNDArray.shape[2],
>           this.inferenceCanvas.width, numRows);
>       await maxValues.data();
> ```

can you put a comment here saying this helps with flickering

---

*[src/math/backends/backend_webgl.ts, line 72 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/466#-L1IYGOulsa65LLZZ6eE:-L1IYGOulsa65LLZZ6eF:bxt58lg) ([raw file](https://github.com/pair-code/deeplearnjs/blob/4249b0e0c75ebc1a3c8483a3f24b30713b1ddfbc/src/math/backends/backend_webgl.ts#L72)):*
> ```XML
>     let texture: WebGLTexture;
>     const texShape: [number, number] = [pixels.height, pixels.width];
>     if (id in this.texData &&
> ```

out of curiosity, when would this be true? in fromPixels, it seems like you make a new NDArray every time

---

*[src/math/backends/backend_webgl.ts, line 784 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/466#-L1IZt1I7HcxKy2wuZxN:-L1IZt1I7HcxKy2wuZxO:bribmnl) ([raw file](https://github.com/pair-code/deeplearnjs/blob/4249b0e0c75ebc1a3c8483a3f24b30713b1ddfbc/src/math/backends/backend_webgl.ts#L784)):*
> ```XML
>     this.texData[id].texture = newTexture;
>     this.texData[id].values = null;
>     if (values != null) {
> ```

when would values be null? should we throw?

---

*[src/math/backends/webgl/texture_manager.ts, line 25 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/466#-L1I_hlZoDzdn-6sTLGn:-L1I_hlZoDzdn-6sTLGo:bf0eug2) ([raw file](https://github.com/pair-code/deeplearnjs/blob/76742f07d27958cfc4af298e4685cbd1ac5aaae7/src/math/backends/webgl/texture_manager.ts#L25)):*
> ```XML
>   private freeTextures: {[shape: string]: WebGLTexture[]} = {};
>   private logEnabled = false;
>   private usedTextureCount: {[shape: string]: number} = {};
> ```

out of curiosity why are you removing this? Because we can move this up into the backend later? I added this for debugging purposes (so we know what shapes are leaking)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/466#-:-L1Ia2mgyMEnhvsfuYCQ:b-rtmsva)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Thanks!

---

Review status: 11 of 14 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[demos/imagenet/imagenet.ts, line 238 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/466#-L1G3ayq7VfQwzyHRyMv:-L1J0EtXIyGpF8polId0:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/4249b0e0c75ebc1a3c8483a3f24b30713b1ddfbc/demos/imagenet/imagenet.ts#L238)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you put a comment here saying this helps with flickering
</blockquote></details>

Done.

---

*[src/math/backends/backend_webgl.ts, line 72 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/466#-L1IYGOulsa65LLZZ6eE:-L1J0KxAeKQmveJeaovp:bwz7lxp) ([raw file](https://github.com/pair-code/deeplearnjs/blob/4249b0e0c75ebc1a3c8483a3f24b30713b1ddfbc/src/math/backends/backend_webgl.ts#L72)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

out of curiosity, when would this be true? in fromPixels, it seems like you make a new NDArray every time
</blockquote></details>

Assume immediate storage mode. Inside `NDArray` c-tor (regardless of whether it's backed by pixels or values), we call `math.write`, which allocates a texture of shape given by `webgl_util.getTexShapeFromLogical`. We need this when we make output array used by gpgpu programs (`this.makeOutputArray`). But then, since we have pixels, we actually want a texShape of `[height, width]` (channels are not part of the shape), and this is achieved by calling `math.writePixels` from `NDArray.fromPixels`,  so we have to dispose the old texture with the incorrect shape.

---

*[src/math/backends/backend_webgl.ts, line 784 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/466#-L1IZt1I7HcxKy2wuZxN:-L1J1T8CLYTU-oG1x5Lv:b8tqzcn) ([raw file](https://github.com/pair-code/deeplearnjs/blob/4249b0e0c75ebc1a3c8483a3f24b30713b1ddfbc/src/math/backends/backend_webgl.ts#L784)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

when would values be null? should we throw?
</blockquote></details>

Values will be null whenever we need to allocate a new array, backed by a new texture, but not populate it with values. We need this when we create an output array for each math operation. See `this.makeOutputArray()`.

---

*[src/math/backends/webgl/texture_manager.ts, line 25 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/466#-L1I_hlZoDzdn-6sTLGn:-L1J3mAI6iyWdzKgYfLj:b-6rt9cz) ([raw file](https://github.com/pair-code/deeplearnjs/blob/76742f07d27958cfc4af298e4685cbd1ac5aaae7/src/math/backends/webgl/texture_manager.ts#L25)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

out of curiosity why are you removing this? Because we can move this up into the backend later? I added this for debugging purposes (so we know what shapes are leaking)
</blockquote></details>

got it. I removed it because it was private and never used to affect logic internally nor exposed externally. Didn't know you are using it in debug mode. Reverting the removal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/466#-:-L1J4We7SxkBK4am7bZk:b-ucfknh)*
<!-- Sent from Reviewable.io -->
